### PR TITLE
[feat] style reset password pages

### PR DESCRIPTION
**Pull request by ccatherinetan**
[//]: # "These comments are meant for your reference. They are invisible and don't need to be deleted!"

### :palm_tree: What's new in this PR
[//]: # "REQUIRED - Describe what's new in this PR in a few lines. A description and bullet points for specifics will suffice."
* created 3 new pages: `src/app/(auth)/reset-password` (styled), `src/app/(auth)/confirm-forgot-password`, `src/app/(auth)/confirm-reset-password`
* added ID functionality to the label of each TextInput component (on login, reset-password, forgot-password pages -- will implement on signup page after that PR is merged)
    * now, if you click on the input label, it will automatically tab into the input box. (shoutout to Jinkang for recommending this). Not sure if there's a naming convention i should follow for id's though.

### :evergreen_tree: Screenshots
[//]: # "REQUIRED for frontend changes, otherwise optional but strongly recommended. Add screenshots of expected behavior - GIFs if you're feeling fancy! If you are making changes here, please CC: @davidqing6432 at the bottom."
Screen Shot 2023-11-07 at 3 48 20 PM
Screen Shot 2023-11-07 at 3 42 56 PM
Screen Shot 2023-11-07 at 3 52 39 PM

## :deciduous_tree: How to review
[//]: # 'REQUIRED - Describe the order in which to review files and what to expect when testing locally. Is there anything specifically you want feedback on? Should this be reviewed commit by commit, or all at once? What are some user flows to test? What are some edge cases to look out for?'
Check routing 
* Forgot Password: http://localhost:3000/forgot-password
    * file`src/app/(auth)/confirm-forgot-password`
    * after clicking "send link to email", check your email for a magic link  (make sure you sign in with your actual email, or one that you have access to!) 
    * click on the link → should route you to the Reset Password Page 
* Reset Password: http://localhost:3000/reset-password
    * file `src/app/(auth)/reset-password` 
    * (after going thru the forgot password flow), complete the reset password flow 
        * to check that your password was reset, sign in again with your new password 
    * you should be signed out after resetting your password  (not sure how to check for this)
    * NOTE: the page will not render unless you access it via the magic link, found in your email 
        *  check that if you reload the page, or visit reset-password again, the page doesn't render. 
* Confirm Reset Password: http://localhost:3000/confirm-reset-password
    * file `src/app/(auth)/confirm-reset-password`
    * click Button → should route to Log In page 

## :seedling: Next steps 
[//]: # "Optional - What's NOT in this PR, doesn't work yet, and/or still needs to be done. Note any temporary fixes in this PR that should be cleaned up later."
* error handling of all inputs (e.g. checking that the new password matches the confirm new password) 
* clearing out the text input fields if there's a reset password error? 

### :information_source: Online sources
[//]: # 'Optional - copy links to any tutorials or documentation that was useful to you when working on this PR'

https://nextjs.org/docs/pages/api-reference/components/link#legacybehavior: for Link usage 

### :potted_plant: Related PRs
[//]: # "Optional - related PRs you're waiting on/ PRs that will conflict, etc; if this is a refactor, feel free to add PRs that previously modified this code"

[//]: # 'This tags the project leader as a default. Feel free to change, or add on anyone who you should be in on the conversation.'
CC: @pragyakallanagoudar
@davidqing6432

### CODE REVIEW

File: src/components/TextInput/index.tsx [hunk 1]
```diff
@@ -4,37 +4,37 @@ import { InputTitleText, InputText, InputDiv, ErrorText } from './styles';
...
 
 export default function TextInput({
   label,
   placeholder,
-  erroring,
   errorText = '',
-  type,
-  name,
+  type = 'text',
+  id, // try to incorporate id? using <label for={id}>
```
Comment (jinkang-0) [inline src/components/TextInput/index.tsx]: Do we still need this comment?

File: src/components/BigButton.tsx [hunk 2]
```diff
@@ -4,9 +4,9 @@ import COLORS from '@/styles/colors';
 const BigButton = styled.button<{ $color?: string }>`
   background-color: ${props => props.color || COLORS.blueMid};
   color: ${props => props.$color};
-  padding: 0.94rem 11.5rem;
+  padding: 0.94rem 0rem;
```
Comment (jinkang-0) [inline src/components/BigButton.tsx]: minor note: for values like 0, we don't need to include the units, so just `0` is fine, but it is up to personal preference!

File: src/app/(auth)/login/page.tsx [hunk 3]
```diff
@@ -56,20 +42,18 @@ export default function Login() {
...
               placeholder="Password"
-              erroring={false}
-              errorText="Password Error"
+              errorText=""
               type="password"
-              name="password"
+              id="password"
               value={password}
               setValue={setPassword}
             />
```
Comment (pragyakallanagoudar) [inline src/app/(auth)/login/page.tsx]: Github does not let me comment outside of the diff, but:

Can we route the "Forgot your password?" text to the `/forgot-password` route instead of `/reset-password`? i.e. on line 62

File: src/components/TextInput/index.tsx (hunk 1 above)
Comment (ccatherinetan) [reply src/components/TextInput/index.tsx]: nope! thanks for catching this!

File: src/components/BigButton.tsx (hunk 2 above)
Comment (ccatherinetan) [reply src/components/BigButton.tsx]: gotcha! changed

File: src/app/(auth)/login/page.tsx (hunk 3 above)
Comment (ccatherinetan) [reply src/app/(auth)/login/page.tsx]: yup!

File: src/app/(auth)/forgot-password/page.tsx [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+        <H1>Forgot Password</H1>
+        <TextInput
+          label="Email"
+          placeholder="email@example.com"
+          type="email"
+          id="email"
+          value={email}
+          setValue={setEmail}
+        />
+        <BigButton type="button" onClick={() => push('/reset-password')}>
```
Comment (pragyakallanagoudar) [inline src/app/(auth)/forgot-password/page.tsx]: Let's try to work the reset password email into this PR! Can you make it so that the button, when pressed, triggers:

```
await supabase.auth.resetPasswordForEmail(email, {
  redirectTo: 'http://localhost:3000/reset-password',
})
```
instead of just routing to `/reset-password`?

See 
* https://supabase.com/docs/guides/auth/auth-password-reset
* https://supabase.com/docs/reference/javascript/auth-resetpasswordforemail

File: src/app/(auth)/reset-password/page.tsx [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+  const [newPassword, setNewPassword] = useState('');
+  const [newPassword2, setNewPassword2] = useState('');
+  const { push } = useRouter();
+  const resetPassword = async () => {
+    const { error } = await supabase.auth.updateUser({ password: newPassword });
+    if (error) {
+      throw new Error(
+        `An error occurred trying to reset password: ${error.message}`,
+      );
+    }
```
Comment (pragyakallanagoudar) [inline src/app/(auth)/reset-password/page.tsx:30]: In addition, can we forcibly log out the user here? i.e. with:

```
const { error } = await supabase.auth.signOut()
```

File: src/app/(auth)/reset-password/page.tsx [hunk 6]
```diff
@@ -0,0 +1,53 @@
+'use client';
```
Comment (pragyakallanagoudar) [inline src/app/(auth)/reset-password/page.tsx:1]: (1) We'll need to make some changes to this to ensure that the page only renders the reset password stuff if we're accessing this route via a magic link that is sent to the user via email. The steps for this are numbered for clarity!

File: src/app/(auth)/reset-password/page.tsx [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+import { useRouter } from 'next/navigation';
+import TextInput from '@/components/TextInput/index';
+import { H1, H4 } from '@/styles/text';
+import { OuterDiv, FormDiv, QuestionsDiv } from '@/app/(auth)/styles';
+import BigButton from '@/components/BigButton';
+import supabase from '../../../api/supabase/createClient';
+
+export default function ResetPassword() {
+  const [newPassword, setNewPassword] = useState('');
+  const [newPassword2, setNewPassword2] = useState('');
```
Comment (pragyakallanagoudar) [inline src/app/(auth)/reset-password/page.tsx:13]: (2) Let's make a new boolean useState called `canReset` or something, and initialize it to false.

File: src/app/(auth)/reset-password/page.tsx [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+  const { push } = useRouter();
```
Comment (pragyakallanagoudar) [inline src/app/(auth)/reset-password/page.tsx:15]: (3) Let's add in a `useEffect()` here that sets the `canReset` useState to true if we consume a `PASSWORD_RECOVERY` event, which can only be triggered by the magic link that is sent via email.

```
useEffect(() => {
   supabase.auth.onAuthStateChange(async (event, session) => {
     if (event == "PASSWORD_RECOVERY") {
       // set the `canReset` useState to true
     }
   })
 }, [])
```

File: src/app/(auth)/reset-password/page.tsx [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+    if (error) {
+      throw new Error(
+        `An error occurred trying to reset password: ${error.message}`,
+      );
+    }
+    push('/confirm-reset-password');
+  };
+
+  return (
+    <OuterDiv>
```
Comment (pragyakallanagoudar) [inline src/app/(auth)/reset-password/page.tsx]: (4) Finally, change this to only render these contents if the `canReset` useState is true.

File: src/app/(auth)/reset-password/page.tsx (hunk 5 above)
Comment (ccatherinetan) [reply src/app/(auth)/reset-password/page.tsx:30]: gotcha, done!

**Review by david-qing — APPROVED** [review]

File: src/app/(auth)/confirm-reset-password/page.tsx [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+import { useRouter } from 'next/navigation';
+import { H1, H4 } from '@/styles/text';
+import { OuterDiv, FormDiv } from '@/app/(auth)/styles';
+import BigButton from '@/components/BigButton';
+
+export default function ConfirmResetPassword() {
+  const { push } = useRouter();
+
+  return (
+    <OuterDiv>
```
Comment (pragyakallanagoudar) [inline src/app/(auth)/confirm-reset-password/page.tsx]: Just one remaining change! Let's get rid of the OuterDiv and FormDiv here since these are already part of the `layout.tsx`.

File: src/app/(auth)/login/page.tsx [hunk 11]
```diff
@@ -49,26 +35,24 @@ export default function Login() {
...
             type="password"
-            name="password"
+            id="password"
             value={password}
             setValue={setPassword}
           />
         </QuestionsDiv>
         <P>
-          <AColored $color={COLORS.greyMid} href="/reset-password">
+          <AColored $color={COLORS.greyMid} href="/forgot-password">
```
Comment (pragyakallanagoudar) [inline src/app/(auth)/login/page.tsx]: Not sure if this is just on my end, but every time I click this for the first time, it renders the `/login` route again 🤔. It's only on the second try that it actually renders `/forgot-password`....

@jinkang-0 curious if you have thoughts on why this might be?

Comment (jinkang-0) [reply src/app/(auth)/login/page.tsx]: I'm not sure if I'm having the same issue... Seems to be redirecting correctly for me.

Comment (pragyakallanagoudar) [reply src/app/(auth)/login/page.tsx]: agh ok it might just be on my end! @ccatherinetan if it looks fine to you too, think we're good :+1:

Comment (jinkang-0) [reply src/app/(auth)/login/page.tsx]: I will add that it may be better if we use the `Link` component from `next/link` to navigate so the page transition is more seamless. More [here](https://nextjs.org/docs/app/api-reference/components/link).

File: src/app/(auth)/reset-password/page.tsx [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+      throw new Error(
+        `An error occurred trying to reset password: ${error.message}`,
+      );
+    }
+    const { error: signOutError } = await supabase.auth.signOut();
+    if (signOutError) {
+      throw new Error(
+        `An error occurred trying to sign out: ${signOutError.message}`,
+      );
+    }
```
Comment (jinkang-0) [inline src/app/(auth)/reset-password/page.tsx:36]: I've been wondering: should we add a `src/api/supabase/auth/...` for supabase auth calls? Could declutter some error handling if we'll be repeating these calls.

File: src/app/(auth)/confirm-reset-password/page.tsx (hunk 10 above)
Comment (ccatherinetan) [reply src/app/(auth)/confirm-reset-password/page.tsx]: gotcha!

File: src/app/(auth)/login/page.tsx (hunk 11 above)
Comment (ccatherinetan) [reply src/app/(auth)/login/page.tsx]: gotcha, i'll implement Link! but otherwise the routing is working fine for me too.

should i use Link for rerouting between other client-side pages as well? @jinkang-0

File: src/app/(auth)/reset-password/page.tsx (hunk 12 above)
Comment (ccatherinetan) [reply src/app/(auth)/reset-password/page.tsx:36]: oooo icic, i'm not entirely sure what this means, but let's discuss today? 

the only supabase i'm making throughout auth is `'@/api/supabase/createClient'`

File: src/app/(auth)/login/page.tsx (hunk 11 above)
Comment (jinkang-0) [reply src/app/(auth)/login/page.tsx]: That'll be great, but just be wary of merge conflicts!

File: src/app/(auth)/login/page.tsx [hunk 13]
```diff
@@ -52,9 +53,9 @@ export default function Login() {
           />
         </QuestionsDiv>
         <P>
-          <AColored $color={COLORS.greyMid} href="/forgot-password">
-            Forgot your password?
-          </AColored>
+          <Link href="/forgot-password" passHref legacyBehavior>
+            <AColored $color={COLORS.greyMid}>Forgot your password?</AColored>
+          </Link>
```
Comment (jinkang-0) [inline src/app/(auth)/login/page.tsx]: When using `Link`, I would suggest passing in the `to` prop instead of an `href`. Using `href` and `legacyBehavior` makes it work the same way as a regular `a` tag. Also, it may be better to turn it into a styled component using `styled(Link)`, perhaps in text styles.

**Review by jinkang-0 — CHANGES_REQUESTED** [review]

**Review by pragyakallanagoudar — APPROVED** [review]

**Review by jinkang-0 — APPROVED** [review]
